Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,695 @@
+    if (!blocks) {
+        return ENOMEM;
+    }
+
+    auto blocks_buffer = UserOrKernelBuffer::for_kernel_buffer(blocks->data());
+    auto did_read = raw_read_blocks(BlockBasedFileSystem::BlockIndex { extent_location }, data_length / logical_block_size(), blocks_buffer);
+    if (!did_read) {
+        return EIO;
+    }
+
+    auto maybe_entry = DirectoryEntry::create(extent_location, data_length, move(blocks));
+    if (maybe_entry.is_error()) {
+        return maybe_entry.error();
+    }
+    m_directory_entry_cache.set(key, maybe_entry.value());
+
+    dbgln_if(ISO9660_DEBUG, "Cached dirent @ {}", extent_location);
+    return maybe_entry.release_value();
+}
+
+u32 ISO9660FS::calculate_directory_entry_cache_key(ISO::DirectoryRecordHeader const* record)
+{
+    return LittleEndian { record->extent_location.little };
+}
+
+KResultOr<size_t> ISO9660Inode::read_bytes(off_t offset, size_t size, UserOrKernelBuffer& buffer, FileDescription*) const
+{
+    MutexLocker inode_locker(m_inode_lock);
+
+    auto& file_system = const_cast<ISO9660FS&>(static_cast<ISO9660FS const&>(fs()));
+    u32 data_length = LittleEndian { m_record.data_length.little };
+    u32 extent_location = LittleEndian { m_record.extent_location.little };
+
+    if (static_cast<u64>(offset) >= data_length)
+        return 0;
+
+    auto block = KBuffer::try_create_with_size(file_system.m_logical_block_size);
+    if (!block) {
+        return ENOMEM;
+    }
+    auto block_buffer = UserOrKernelBuffer::for_kernel_buffer(block->data());
+
+    size_t total_bytes = min(size, data_length - offset);
+    size_t nread = 0;
+    size_t blocks_already_read = offset / file_system.m_logical_block_size;
+    size_t initial_offset = offset % file_system.m_logical_block_size;
+
+    auto current_block_index = BlockBasedFileSystem::BlockIndex { extent_location + blocks_already_read };
+    while (nread != total_bytes) {
+        size_t bytes_to_read = min(total_bytes - nread, file_system.logical_block_size() - initial_offset);
+        auto buffer_offset = buffer.offset(nread);
+        dbgln_if(ISO9660_VERY_DEBUG, "ISO9660Inode::read_bytes: Reading {} bytes into buffer offset {}/{}, logical block index: {}", bytes_to_read, nread, total_bytes, current_block_index.value());
+
+        if (bool result = file_system.raw_read(current_block_index, block_buffer); !result) {
+            return EIO;
+        }
+
+        bool result = buffer_offset.write(block->data() + initial_offset, bytes_to_read);
+        if (!result) {
+            return EFAULT;
+        }
+
+        nread += bytes_to_read;
+        initial_offset = 0;
+        current_block_index = BlockBasedFileSystem::BlockIndex { current_block_index.value() + 1 };
+    }
+
+    return nread;
+}
+
+InodeMetadata ISO9660Inode::metadata() const
+{
+    return m_metadata;
+}
+
+KResult ISO9660Inode::traverse_as_directory(Function<bool(FileSystem::DirectoryEntryView const&)> visitor) const
```

Comment:
Is GitHub search failing me.. or does no-one call this function?

Reply:
This is a virtual override, `VirtualFileSystem` calls this.